### PR TITLE
Remove the need for rust-mozjs to use unstable Rust features

### PR DESCRIPTION
**Pull request by SimonSapin**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18875)

### CODE REVIEW

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 1]
```diff
@@ -3660,18 +3661,16 @@ def jitInfoInitializer(isTypedMethod):
...
-                            JSValueType::${returnType} as u8,
-                            ${isInfallible},
-                            ${isMovable},
-                            ${isEliminatable},
-                            ${isAlwaysInSlot},
-                            ${isLazilyCachedInSlot},
-                            ${isTypedMethod},
-                            ${slotIndex} as u16,
-                        )
+                        ((JSJitInfo_OpType::${opType} as u8 as u32) << 0) |
```
Comment (SimonSapin) [inline components/script/dom/bindings/codegen/CodegenRust.py]: Next I’ll move this bit-shifting into a macro in rust-mozjs

**Comment by nox** [comment]
I'm not really ok with the removal of `NonZero` without any proof that it could break in the future. We can just make a `NonZeroPtr<T>` that stores a `&'static T` or something. I'll defer to @jdm for the reviewal.

The rest looks fine to me, but it sounds like you want to remove all `const` functions from `script`, which I don't think is a good idea either, given that would require quite a bit of changes that will make the code way more uglier.

**Comment by SimonSapin** [comment]
The default is still to use `core::nonzero::NonZero` with its memory layout optimization. This only changes memory layout when compiling with `--no-default-features --features default-except-unstable`. And I don’t expect anyone to actually use a build compiled with this mode. It only exists to enable more parts of Servo to be included in Rust’s upstream regression testing and performance testing.

I’m not planning to touch `const fn`’s in script. The `JSJitInfo::new_bitfield_1` change is comparatively easy, and allows limiting the required unstable features to the servo/servo repo.

**Comment by SimonSapin** [comment]
Pushed a few more commits. The last one updates rust-mozjs to a commit that hasn’t been merged yet: https://github.com/servo/rust-mozjs/pull/373 so this PR should not land until that one has been reviewed.

**Comment by jdm** [comment]
This looks fine to me.

**Pull request merged**